tests/integration_tests/test_subtensor_integration.py

```diff
@@ -1,407 +1,143 @@
-                side_effect=is_registered_return_values
-            )
-
-            self.subtensor.get_current_block = MagicMock(side_effect=current_block)
-            self.subtensor.get_neuron_for_pubkey_and_subnet = MagicMock(
-                return_value=mock_neuron
-            )
-            self.subtensor.substrate.get_block_hash = MagicMock(
-                return_value="0x" + "0" * 64
-            )
-            self.subtensor._do_pow_register = MagicMock(return_value=(False, "Failed"))
-
-            # should return True
-            self.assertIsNot(
-                self.subtensor.register(wallet=wallet, netuid=3),
-                True,
-                msg="Registration should fail",
-            )
-            self.assertEqual(mock_create_pow.call_count, 3)
-
-    def test_registration_stale_then_continue(self):
-        # verify that after a stale solution, to solve will continue without exiting
-
-        class ExitEarly(Exception):
-            pass
-
-        mock_is_stale = MagicMock(side_effect=[True, False])
-
-        mock_do_pow_register = MagicMock(side_effect=ExitEarly())
-
-        mock_subtensor_self = MagicMock(
-            neuron_for_pubkey=MagicMock(
-                return_value=MagicMock(is_null=True)
-            ),  # not registered
-            substrate=MagicMock(
-                get_block_hash=MagicMock(return_value="0x" + "0" * 64),
-            ),
-        )
-
-        mock_wallet = MagicMock()
-
-        mock_create_pow = MagicMock(return_value=MagicMock(is_stale=mock_is_stale))
-
-        with patch(
-            "bittensor.core.extrinsics.registration.create_pow", mock_create_pow
-        ), patch(
-            "bittensor.core.extrinsics.registration._do_pow_register",
-            mock_do_pow_register,
-        ):
-            # should create a pow and check if it is stale
-            # then should create a new pow and check if it is stale
-            # then should enter substrate and exit early because of test
-            self.subtensor.get_neuron_for_pubkey_and_subnet = MagicMock(
-                return_value=bittensor.NeuronInfo.get_null_neuron()
-            )
-            with pytest.raises(ExitEarly):
-                bittensor.subtensor.register(mock_subtensor_self, mock_wallet, netuid=3)
-            self.assertEqual(
-                mock_create_pow.call_count, 2, msg="must try another pow after stale"
-            )
-            self.assertEqual(mock_is_stale.call_count, 2)
-            self.assertEqual(
-                mock_do_pow_register.call_count,
-                1,
-                msg="only tries to submit once, then exits",
-            )
-
-
-if __name__ == "__main__":
-    unittest.main()
+from bittensor.core.chain_data.axon_info import AxonInfo
+
+from bittensor import NeuronInfo
+from bittensor.core.subtensor import Subtensor
+from tests.helpers.integration_websocket_data import WEBSOCKET_RESPONSES
+
+
+class FakeWebsocket(ClientConnection):
```

Comment:
let's consider keep this class in some helper module if we gonna use in different test modules.

Reply:
Good idea